### PR TITLE
fix class names in formatter

### PR DESCRIPTION
**Pull request by smirnovaka**

### CODE REVIEW

File: src/foam/core/AbstractFObjectPropertyInfo.java [hunk 1]
```diff
@@ -127,7 +128,13 @@ public boolean isDefaultValue(Object o) {
   }
   
   public void format(foam.lib.formatter.FObjectFormatter formatter, foam.core.FObject obj) {
-    formatter.output(get_(obj));
+    Object propObj = get_(obj);
+    if ( propObj instanceof FObject && ! (propObj instanceof UnknownFObject) ) {
```
Comment (kgrgreer) [inline src/foam/core/AbstractFObjectPropertyInfo.java]: When is it not an FObject? When it is null?

File: src/foam/lib/formatter/JSONFObjectFormatter.java [hunk 2]
```diff
@@ -86,7 +86,7 @@ public JSONFObjectFormatter() {
     super();
   }
 
-  protected void outputUndefined() {
+  protected void outpvutUndefined() {
```
Comment (kgrgreer) [inline src/foam/lib/formatter/JSONFObjectFormatter.java]: typo "output" not "outpvut"

File: src/foam/lib/formatter/JSONFObjectFormatter.java [hunk 3]
```diff
@@ -74,7 +74,7 @@ protected SimpleDateFormat initialValue() {
   protected boolean outputShortNames_    = true;
   protected boolean outputDefaultValues_ = false;
   protected boolean multiLineOutput_     = false;
-  protected boolean outputClassNames_    = true;
+//  protected boolean outputClassNames_    = true;
```
Comment (kgrgreer) [inline src/foam/lib/formatter/JSONFObjectFormatter.java]: Why is this commented out?

File: src/foam/core/AbstractFObjectPropertyInfo.java (hunk 1 above)
Comment (smirnovaka) [reply src/foam/core/AbstractFObjectPropertyInfo.java]: in cases like this:
```{
     class: 'FObjectProperty',
     javaType: 'java.util.Timer',
     name: 'timer',
     javaFactory: 'return new Timer();',
   }

File: src/foam/lib/formatter/JSONFObjectFormatter.java (hunk 3 above)
Comment (smirnovaka) [reply src/foam/lib/formatter/JSONFObjectFormatter.java]: it's not used. Removed.

File: src/foam/core/AbstractFObjectPropertyInfo.java (hunk 1 above)
Comment (kgrgreer) [reply src/foam/core/AbstractFObjectPropertyInfo.java]: Then why is it an FObjectProperty? This seems wrong. If it isn't an FObject then it shouldn't be an FObjectProperty. java.util.Timer isn't going to be serializable anyway.

File: src/foam/lib/formatter/JSONFObjectFormatter.java [hunk 4]
```diff
@@ -74,7 +73,6 @@ protected SimpleDateFormat initialValue() {
   protected boolean outputShortNames_    = true;
   protected boolean outputDefaultValues_ = false;
   protected boolean multiLineOutput_     = false;
-  protected boolean outputClassNames_    = true;
```
Comment (kgrgreer) [inline src/foam/lib/formatter/JSONFObjectFormatter.java:77]: Why is this removed?

File: src/foam/lib/formatter/JSONFObjectFormatter.java [hunk 5]
```diff
@@ -303,7 +302,16 @@ protected Boolean maybeOutputProperty(FObject fo, PropertyInfo prop, boolean inc
...
-    ClassInfo info           = oldFObject.getClassInfo();
+    outputDelta(oldFObject, newFObject, null);
+  }
+
+  public void outputDelta(FObject oldFObject, FObject newFObject, ClassInfo defaultClass) {
+    ClassInfo info = newFObject.getClassInfo();
+
+    boolean outputClass = false;
+
+    if ( outputClassNames_ && ( info != defaultClass || outputDefaultClassNames_ ) ) outputClass = true;
```
Comment (kgrgreer) [inline src/foam/lib/formatter/JSONFObjectFormatter.java]: You could combine this line and the one above:
boolean outputClass = outputClassNames_ && ( info != defaultClass || outputDefaultClassNames_ );

File: src/foam/lib/formatter/JSONFObjectFormatter.java [hunk 6]
```diff
@@ -370,25 +370,18 @@ public void output(FObject[] arr) {
...
 
-    if ( info == defaultClass ) outputDefaultClassNames_ = false;
+    boolean outputClass = false;
 
-    output(o);
-  }
-
-  public void output(FObject o) {
-    ClassInfo info = o.getClassInfo();
+    if ( outputClassNames_ && ( info != defaultClass || outputDefaultClassNames_ ) ) outputClass = true;
```
Comment (kgrgreer) [inline src/foam/lib/formatter/JSONFObjectFormatter.java]: same as above comment
